google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceCleaner.java

```diff
@@ -0,0 +1,210 @@
+      this.res = res;
+
+      @SuppressWarnings("unchecked")
+      Function<T, Void> localFunc = (Function<T, Void>) getDeleteMethod(res);
+
+      this.func = localFunc;
+    }
+
+    public void delete() {
+      func.apply(res);
+    }
+  }
+
+  public void cleanUp() {
+    for (Map.Entry<Object, ManagedResource<?>> r : resources.entrySet()) {
+      System.out.println("Clean up compute engine resource: " + r.getKey().toString());
+      r.getValue().delete();
+    }
+    resources.clear();
+  }
+
+  public ManagedResourceCleaner add(AddressId addressId) {
+    ManagedResource<AddressId> r = new ManagedResource<>(addressId);
+    resources.put(addressId, r);
+    return this;
+  }
+
+  public ManagedResourceCleaner remove(AddressId addressId) {
+    resources.remove(addressId);
+    return this;
+  }
+
+  public ManagedResourceCleaner add(DiskId diskId) {
+    ManagedResource<DiskId> r = new ManagedResource<>(diskId);
+    resources.put(diskId, r);
+    return this;
+  }
+
+  public ManagedResourceCleaner remove(DiskId diskId) {
+    resources.remove(diskId);
+    return this;
+  }
+
+  public ManagedResourceCleaner add(SnapshotId snapshotId) {
+    ManagedResource<SnapshotId> r = new ManagedResource<>(snapshotId);
+    resources.put(snapshotId, r);
+    return this;
+  }
+
+  public ManagedResourceCleaner remove(SnapshotId snapshotId) {
+    resources.remove(snapshotId);
+    return this;
+  }
+
+  public ManagedResourceCleaner add(NetworkId networkId) {
+    ManagedResource<NetworkId> r = new ManagedResource<>(networkId);
+    resources.put(networkId, r);
+    return this;
+  }
+
+  public ManagedResourceCleaner remove(NetworkId networkId) {
+    resources.remove(networkId);
+    return this;
+  }
+
+  public ManagedResourceCleaner add(SubnetworkId subnetworkId) {
+    ManagedResource<SubnetworkId> r = new ManagedResource<>(subnetworkId);
+    resources.put(subnetworkId, r);
+    return this;
+  }
+
+  public ManagedResourceCleaner remove(SubnetworkId subnetworkId) {
+    resources.remove(subnetworkId);
+    return this;
+  }
+
+  public ManagedResourceCleaner add(ImageId imageId) {
+    ManagedResource<ImageId> r = new ManagedResource<>(imageId);
+    resources.put(imageId, r);
+    return this;
+  }
+
+  public ManagedResourceCleaner remove(ImageId imageId) {
+    resources.remove(imageId);
+    return this;
+  }
+
+  public ManagedResourceCleaner add(InstanceId instanceId) {
+    ManagedResource<InstanceId> r = new ManagedResource<>(instanceId);
+    resources.put(instanceId, r);
+    return this;
+  }
+
+  public ManagedResourceCleaner remove(InstanceId instanceId) {
+    resources.remove(instanceId);
+    return this;
+  }
+
+  private <T> Function<?, Void> getDeleteMethod(T res) {
+
+    if (res instanceof AddressId) {
+      return new Function<AddressId, Void>() {
+        @Override
+        public Void apply(AddressId addressId) {
+          compute.deleteAddress(addressId);
+          return null;
+        }
+      };
```

Comment:
It would be much cleaner to put these delete `Function`s inside the corresponding `add` method for each type. Then, you never have to bother with an `instanceof` check. Additionally, I think it would mean you could remove the type parameterization (`<T>`) from MonitoredResource too.